File: packages/node/src/mcp/validate-tool-input.ts
import type { SupportedLang } from '../i18n/lang.js';
import { buildToolDefinitions } from './tools.js';

/** Returns `null` if valid; otherwise a human-readable message for HTTP 400. */
export function validateToolInput(name: string, args: Record<string, unknown>, locale: SupportedLang): string | null {
  const defs = buildToolDefinitions(locale);
  const schema =
    name === 'ping'
      ? defs.ping.schema
      : name === 'scrape'
        ? defs.scrape.schema
        : name === 'scrape_article'
          ? defs.scrape_article.schema
          : name === 'scrape_emails'
            ? defs.scrape_emails.schema
            : name === 'scrape_phones'
              ? defs.scrape_phones.schema
              : name === 'scrape_urls'
                ? defs.scrape_urls.schema
                : name === 'scrape_images'
                  ? defs.scrape_images.schema
                  : null;
  if (!schema) return 'Unknown tool name';
  const parsed = schema.safeParse(args);
  if (parsed.success) return null;
  return parsed.error.issues.map((i) => `${i.path.length ? i.path.join('.') : 'root'}: ${i.message}`).join('; ');
}

File: packages/python/src/lionscraper/mcp/validate_tool_input.py
from __future__ import annotations

from typing import Any

from pydantic import ValidationError

from lionscraper.i18n.lang import SupportedLang
from lionscraper.mcp.tools import build_tool_definitions


def validate_tool_input(name: str, args: dict[str, Any], locale: SupportedLang) -> str | None:
    defs = build_tool_definitions(locale)
    entry = defs.get(name)
    if not entry:
        return "Unknown tool name"
    model = entry["model"]
    try:
        model.model_validate(args)
        return None
    except ValidationError as e:
        parts: list[str] = []
        for err in e.errors():
            loc = err.get("loc") or ()
            path = ".".join(str(x) for x in loc) if loc else "root"
            parts.append(f"{path}: {err.get('msg', 'invalid')}")
        return "; ".join(parts)
